File: remix/src/core/path-match.ts
/**
 * Glob-style path matcher with `*` wildcard. `*` matches any character
 * (including `/`) so `/admin/*` matches `/admin/foo/bar`.
 */

const SPECIAL_CHARS = /[.+?^${}()|[\]\\]/g;

export function matchPath(pathname: string, pattern: string): boolean {
  if (!pattern) return false;
  if (pattern === pathname) return true;
  if (!pattern.includes('*')) return false;

  const escaped = pattern.replace(SPECIAL_CHARS, '\\$&').replace(/\*/g, '.*');
  const regex = new RegExp(`^${escaped}$`);
  return regex.test(pathname);
}

export function isPathProtected(pathname: string, patterns: string[] = []): boolean {
  if (!patterns.length) return false;
  return patterns.some((pattern) => matchPath(pathname, pattern));
}

export function isPathExcluded(pathname: string, patterns: string[] = []): boolean {
  if (!patterns.length) return false;
  return patterns.some((pattern) => matchPath(pathname, pattern));
}

File: rails/lib/showad/path_match.rb
# frozen_string_literal: true

module ShowAd
  # Glob-style path matching shared by the middleware's `protected_paths` and
  # `excluded_paths` configuration. Supports a single `*` wildcard which
  # matches any character including `/` (matches the Laravel and Next.js
  # SDKs' behavior).
  module PathMatch
    module_function

    # @param path    [String] request path, leading slash optional
    # @param pattern [String] pattern, may contain `*`
    def matches?(path, pattern)
      return false if path.nil? || pattern.nil?

      normalized_path = '/' + path.to_s.sub(%r{\A/+}, '')
      normalized_pattern = '/' + pattern.to_s.sub(%r{\A/+}, '')

      return true if normalized_path == normalized_pattern

      if normalized_pattern.include?('*')
        regex = Regexp.escape(normalized_pattern).gsub('\\*', '.*')
        return Regexp.new("\\A#{regex}\\z").match?(normalized_path)
      end

      false
    end

    # @return [Boolean] true when `path` matches at least one pattern.
    def matches_any?(path, patterns)
      Array(patterns).any? { |p| matches?(path, p) }
    end
  end
end
